src/backend/discovery/discovery.lambda.ts

```diff
@@ -0,0 +1,297 @@
+import * as console from 'console';
+import * as https from 'https';
+import * as process from 'process';
+import { URL } from 'url';
+
+// eslint-disable-next-line import/no-unresolved
+import type { Context } from 'aws-lambda';
+import * as AWS from 'aws-sdk';
+import * as Nano from 'nano';
+import { IngestionInput, integrity } from '../shared';
+
+const TIMEOUT_MILLISECONDS = 10_000;
+const CONSTRUCT_KEYWORDS: ReadonlySet<string> = new Set(['cdk', 'aws-cdk', 'cdk8s', 'cdktf']);
+const MARKER_FILE_NAME = 'couchdb-last-transaction-id';
+const NPM_REPLICA_REGISTRY_URL = 'https://replicate.npmjs.com/';
+
+/**
+ * The S3 prefix for storing change that we failed processing
+ */
+const FAILED_CHANGE_PREFIX = 'failed';
+
+let s3: AWS.S3 | undefined;
+let sqs: AWS.SQS | undefined;
+
+let stagingBucket: string;
+let queueUrl: string;
+
+/**
+ * This function triggers on a fixed schedule and reads a stream of changes frm npmjs couchdb _changes endpoint.
+ * Upon invocation the function starts reading from a sequence stored in an s3 object - the `marker`.
+ * If the marker fails to load (or do not exist), the stream will start from `now` - the latest change.
+ * For each change:
+ *  - the package version tarball will be copied from the npm registry to a stating bucket.
+ *  - a message will be sent to an sqs queue
+ * Currently we don't handle the function execution timeout, and accept that the last batch processed might be processed again,
+ * relying on the idempotency on the consumer side.
+ * npm registry API docs: https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md
+ * @param context a Lambda execution context
+ */
+export async function handler(_request: unknown, context: Context) {
+  if (s3 == null) {
+    s3 = new AWS.S3();
+  }
+  if (sqs == null) {
+    sqs = new AWS.SQS();
+  }
+
+  if (!process.env.STAGING_BUCKET_NAME) {
+    throw new Error('The STAGING_BUCKET_NAME environment variable is not set');
+  }
+  stagingBucket = process.env.STAGING_BUCKET_NAME;
+
+  if (!process.env.QUEUE_URL) {
+    throw new Error('The QUEUE_URL environment variable is not set');
+  }
+  queueUrl = process.env.QUEUE_URL;
+
+  // load marker from the staging bucket
+  let marker;
+  try {
+    await s3.getObject({
+      Bucket: stagingBucket,
+      Key: MARKER_FILE_NAME,
+    }).promise().then(function(data) {
+      marker = data.Body?.toString('utf-8');
+    });
```

Comment:
It's not a good practice to mix async with then():

```suggestion
    marker = (await s3.getObject({
      Bucket: stagingBucket,
      Key: MARKER_FILE_NAME,
    }).promise()).Body?.toString('utf-8');
    });
```